src/interface.jl

```diff
@@ -31,111 +46,201 @@ ADBackend(::Val) = error("The requested AD backend is not available. Make sure t
-abstract type Bijector{N} <: AbstractBijector end
+Transform `x` using `b`, storing the result in `y`.
 
-dimension(b::Bijector{N}) where {N} = N
-dimension(b::Type{<:Bijector{N}}) where {N} = N
+If `y` is not provided, `x` is used as the output.
+"""
+transform!(b, x) = transform!(b, x, x)
+transform!(b, x, y) = (y .= transform(b, x))
 
-Broadcast.broadcastable(b::Bijector) = Ref(b)
+"""
+    logabsdetjac(b, x)
 
+Return `log(abs(det(J(b, x))))`, where `J(b, x)` is the jacobian of `b` at `x`.
 """
-    isclosedform(b::Bijector)::bool
-    isclosedform(b⁻¹::Inverse{<:Bijector})::bool
+logabsdetjac(b, x) = last(with_logabsdet_jacobian(b, x))
+
+"""
+    logabsdetjac!(b, x[, logjac])
+
+Compute `log(abs(det(J(b, x))))` and store the result in `logjac`, where `J(b, x)` is the jacobian of `b` at `x`.
+"""
+logabsdetjac!(b, x) = logabsdetjac!(b, x, zero(eltype(x)))
+logabsdetjac!(b, x, logjac) = (logjac += logabsdetjac(b, x))
+
+"""
+    with_logabsdet_jacobian!(b, x[, y, logjac])
+
+Compute `transform(b, x)` and `logabsdetjac(b, x)`, storing the result
+in `y` and `logjac`, respetively.
+
+If `y` is not provided, then `x` will be used in its place.
+
+Defaults to calling `with_logabsdet_jacobian(b, x)` and updating `y` and `logjac` with the result.
+"""
+with_logabsdet_jacobian!(b, x) = with_logabsdet_jacobian!(b, x, x)
+with_logabsdet_jacobian!(b, x, y) = with_logabsdet_jacobian!(b, x, y, zero(eltype(x)))
+function with_logabsdet_jacobian!(b, x, y, logjac)
+    y_, logjac_ = with_logabsdet_jacobian(b, x)
+    y .= y_
+    return (y, logjac + logjac_)
+end
+
+"""
+    isclosedform(b::Transform)::bool
+    isclosedform(b⁻¹::Inverse{<:Transform})::bool
 
 Returns `true` or `false` depending on whether or not evaluation of `b`
 has a closed-form implementation.
 
-Most bijectors have closed-form evaluations, but there are cases where
+Most transformations have closed-form evaluations, but there are cases where
 this is not the case. For example the *inverse* evaluation of `PlanarLayer`
 requires an iterative procedure to evaluate.
 """
-isclosedform(b::Bijector) = true
+isclosedform(t::Transform) = true
 
+# Invertibility "trait".
 """
-    inverse(b::Bijector)
-    Inverse(b::Bijector)
+    Invertible
 
-A `Bijector` representing the inverse transform of `b`.
+Represents the trait of being, well, non-invertible.
+"""
+struct NotInvertible end
 """
-struct Inverse{B <: Bijector, N} <: Bijector{N}
-    orig::B
+    Invertible
 
-    Inverse(b::B) where {N, B<:Bijector{N}} = new{B, N}(b)
-end
+Represents the trait of being, well, invertible.
+"""
+struct Invertible end
 
-# field contains nested numerical parameters
-Functors.@functor Inverse
+# Useful for checking if compositions, etc. are invertible or not.
+Base.:+(::NotInvertible, ::Invertible) = NotInvertible()
+Base.:+(::Invertible, ::NotInvertible) = NotInvertible()
+Base.:+(::NotInvertible, ::NotInvertible) = NotInvertible()
+Base.:+(::Invertible, ::Invertible) = Invertible()
```

Comment:
I would recommend removing these definitions. Seems like a misuse of `+`.

Reply:
Done!